### PR TITLE
Enrich Dex logs with real IP and request ID

### PR DESCRIPTION
**Pull request by nabokihms**
#### Overview

- Include client's real IP in logs.
- Capture and log unique request IDs.
- Update middleware to extract and forward this data.
- Enhance debugging and traceability. Improves insight and troubleshooting efficiency.

#### What this PR does / why we need it

https://github.com/dexidp/dex/issues/1788

### CODE REVIEW

File: cmd/dex/config.go [hunk 1]
```diff
@@ -182,15 +183,36 @@ type OAuth2 struct {
...
+	TLSMinVersion  string         `json:"tlsMinVersion"`
+	TLSMaxVersion  string         `json:"tlsMaxVersion"`
+	AllowedOrigins []string       `json:"allowedOrigins"`
+	AllowedHeaders []string       `json:"allowedHeaders"`
+	ClientRemoteIP ClientRemoteIP `json:"clientRemoteIP"`
+}
+
+type ClientRemoteIP struct {
+	Header            string   `json:"header"`
+	TrustedProxyCIDRs []string `json:"trustedProxyCIDRs"`
```
Comment (sagikazarmark) [inline cmd/dex/config.go:200]: How about just calling this `TrustedProxies`? Sounds simpler and I think people will still understand they need to pass CIDRs in there.

File: cmd/dex/config.go [hunk 2]
```diff
@@ -182,15 +183,36 @@ type OAuth2 struct {
...
+}
+
+func (cr *ClientRemoteIP) ToParsedCIDRs() (string, []netip.Prefix, error) {
+	if cr == nil {
+		return "", nil, nil
+	}
```
Comment (sagikazarmark) [inline cmd/dex/config.go:206]:
```suggestion
	}

```

Note to self: add more linter rules.

File: cmd/dex/config.go [hunk 3]
```diff
@@ -182,15 +183,36 @@ type OAuth2 struct {
...
+	trusted := make([]netip.Prefix, 0, len(cr.TrustedProxyCIDRs))
+	for _, cidr := range cr.TrustedProxyCIDRs {
+		ipNet, err := netip.ParsePrefix(cidr)
+		if err != nil {
+			return "", nil, fmt.Errorf("failed to parse CIDR %q: %v", cidr, err)
+		}
+		trusted = append(trusted, ipNet)
+	}
```
Comment (sagikazarmark) [inline cmd/dex/config.go:214]:
```suggestion
	}

```

File: cmd/dex/config.go [hunk 4]
```diff
@@ -182,15 +183,36 @@ type OAuth2 struct {
...
+	AllowedHeaders []string       `json:"allowedHeaders"`
+	ClientRemoteIP ClientRemoteIP `json:"clientRemoteIP"`
+}
+
+type ClientRemoteIP struct {
+	Header            string   `json:"header"`
+	TrustedProxyCIDRs []string `json:"trustedProxyCIDRs"`
+}
+
+func (cr *ClientRemoteIP) ToParsedCIDRs() (string, []netip.Prefix, error) {
```
Comment (sagikazarmark) [inline cmd/dex/config.go:203]: Why return the header here?

File: server/server.go [hunk 5]
```diff
@@ -84,6 +87,9 @@ type Config struct {
 
 	// Headers is a map of headers to be added to the all responses.
 	Headers http.Header
+	// Header to extract real ip from.
```
Comment (sagikazarmark) [inline server/server.go:90]:
```suggestion

	// Header to extract real ip from.
```

File: server/server.go [hunk 6]
```diff
@@ -682,3 +729,18 @@ func (s *Server) getConnector(id string) (Connector, error) {
 
 	return conn, nil
 }
+
+type logRequestKey string
+
+const (
+	LogRequestKeyRequestID logRequestKey = "request_id"
+	LogRequestKeyRemoteIP  logRequestKey = "client_remote_addr"
```
Comment (sagikazarmark) [inline server/server.go:737]:
```suggestion
	RequestKeyRequestID logRequestKey = "request_id"
	RequestKeyRemoteIP  logRequestKey = "client_remote_addr"
```

Drop the log part: we don't use it for anything else, but at the point of insertion it doesn't really matter IMO.

File: server/server.go [hunk 7]
```diff
@@ -682,3 +729,18 @@ func (s *Server) getConnector(id string) (Connector, error) {
...
+)
+
+func SetLogRequestID(ctx context.Context) context.Context {
```
Comment (sagikazarmark) [inline server/server.go:740]:
```suggestion
func WithRequestID(ctx context.Context) context.Context {
```

**Review by sagikazarmark — COMMENTED** [review]
LGTM, couple suggestions

File: cmd/dex/config.go (hunk 1 above)
Comment (nabokihms) [reply cmd/dex/config.go]: Ok, it does make sense. Fixed.

File: cmd/dex/config.go (hunk 4 above)
Comment (nabokihms) [reply cmd/dex/config.go]: At first, I thought this method could prepare everything for IP address detection. If you think this is odd, I can just remove it, if you don't mind. 

Fixed.

**Review by sagikazarmark — APPROVED** [review]

File: cmd/dex/config.go (hunk 4 above)
Comment (sagikazarmark) [reply cmd/dex/config.go]: I think I like the latest version better. If there is need for some reason to pack all that into a single method, I'd call it something like `prepareClientIPDetection`.

But as I said, I like the latest version.

**Pull request merged**